Improvements (1.21.11)

litetex opened the pull request:
Contains various fixes and improvements and updates the mod for 1.21.11:
* Updated gradle
* Improve project setup/templating
* Fix and improve rendering of options screen
  * Now responsive
  * Requires less rendering work
  * Rendered correctly on 1.21.11
  * Fix memory leak due to `currentSoundInstance` being not cleared after leaving the options screen
  * Fix sounds keeping playing after leaving the screen
* Correctly mixin to `SoundOptionsScreen`
* Improve code formatting in some places
* Improve file saving
* Removed old configuration migrations as they are probably outdated by now

litetex committed:
Update gradle

litetex commented:
@BVengo Updated as requested.

I also fixed 2 additional problems:
1. memory leak due to `currentSoundInstance` being not cleared after leaving the options screen
2. sounds keeping playing after leaving the options screen

BVengo commented on gradle.properties:
```diff
@@ -3,19 +3,17 @@ org.gradle.jvmargs=-Xmx1G
 
 # Mod Properties
-mod_version = 1.2.4
+mod_version=1.3.0
```
I'll be bumping as a minor update, since it's only bug fixes (from the user perspective).

```suggestion
mod_version=1.2.5
```

BVengo commented on src/main/resources/fabric.mod.json:
```diff
@@ -27,21 +29,21 @@
+        "minecraft": ">=${minecraft_version}",
+        "fabric-api": ">=${fabric_api_version}",
+        "modmenu": ">=${modmenu_version}"
     },
```
Interesting - why the switch from depends to recommends here?

BVengo commented on src/main/java/com/bvengo/soundcontroller/config/ConfigParser.java:
I'll be retaining the backwards compatibility. It doesn't slow things down much since we're doing a version check anyway, and I know that people will complain when old modpacks update to newer versions :)

BVengo commented on src/main/java/com/bvengo/soundcontroller/gui/VolumeWidgetEntry.java:
```diff
@@ -88,59 +72,87 @@ private void addSlider() {
+                    ((ResettableOptionWidget) this.volumeSlider).resetValue();
+                    this.updateElementWidths();
+                    this.updateElementPositions();
+                });
```
Why does this need to resize / reposition elements when reset? It should just be resetting the volume data which shouldn't have any positional impact.

BVengo reviewed (changes requested):
Thanks so much litetex, it looks great.

I've added a few comments (mostly clarifications for my own understanding, but some changes). Could you please check them out, and then I'll see if we can get this merged in :+1:

litetex commented on src/main/java/com/bvengo/soundcontroller/SoundController.java:
```diff
@@ -26,8 +26,16 @@ public void onInitializeClient() {
+							}
+
+							CONFIG.reload();
+						});
```
Well the external file was quite big, but yeah when I now think about it, it might have been the better way.

I'm however not sure why this is even needed in the first place...
The changelog of [1.2.4](https://github.com/BVengo/sound-controller/releases/tag/v1.2.4) states:

> This is useful not just for when you add new resource packs, but also when you modify your config file directly.

1. AFAIK resource packs don't add new sounds
2. Why slow down the loading of resource packs? Why don't just add a button in the UI that says "Reload config from disk" or something like this? I'm also not sure why someone would manually edit the config file in the first place while the game is running...

BVengo replied:
1. They can add new sounds (datapacks too), used heavily in customised servers such as Hypixel, MCC, etc. I've had a lot of reports of them not properly loading, and this change fixed all issues by ensuring that sounds are re-checked during the refresh (and new ones added).

2. All the users who overload the value, but also useful for debugging. This wasn't the primary purpose, just a happy accident.

BVengo replied:
1. They can add new sounds (datapacks too), used heavily in customised servers such as Hypixel, MCC, etc. I've had a lot of reports of them not properly loading, and this change fixed all issues by ensuring that sounds are re-checked during the refresh (and new ones added).

2. All the users who overload the value, but also useful for debugging. This wasn't the primary purpose, just a happy accident.

litetex commented:
@BVengo 
Thank you for the detailed review. I addressed all problems and some require further discussion. Could you have a look again? :)

litetex committed:
Various improvements

[SQUASHED] Ported from previous version as there were changes done in the meantime. See https://github.com/BVengo/sound-controller/pull/33

* Improve project setup/templating
* Fix and improve rendering of options screen
  * Now responsive
  * Requires less rendering work
  * Rendered correctly on 1.21.11
  * Fix memory leak due to `currentSoundInstance`
  * Fix sounds keeping playing after leaving the screen
* Correctly mixin to `SoundOptionsScreen`
* Improve code formatting in some places
* Improve file saving
* Removed old configuration migrations as they are probably outdated by now

Restore sound reload listener

Improve parsing of config - Readd old logic

Apparently a ton of people are still using 1.20.1 -.-

This is no longer required as the slider is no longer re-created when clicking reset

As requested: Only do patch version bump

Cleanup, call onClose from upstream

litetex committed:
Cleanup
